components/collector/src/source_collectors/cargo_audit/security_warnings.py

```diff
@@ -1,32 +1,78 @@
 
-from typing import cast
+from typing import Literal, TypedDict, cast
 
 from base_collectors import JSONFileSourceCollector
-from collector_utilities.type import JSON, JSONDict
+from collector_utilities.type import JSON
 from model import Entities, Entity
 
 
+class Advisory(TypedDict):
+    """A Cargo audit advisory."""
+
+    id: str
+    title: str
+    url: str
+
+
+class Package(TypedDict):
+    """A Cargo audit package description."""
+
+    name: str
+    version: str
+
+
+class CargoAuditWarning(TypedDict):
+    """A Cargo audit warning."""
+
+    advisory: Advisory
+    package: Package
+    versions: dict[Literal["patched"], list[str]]
+    kind: Literal["vulnerability", "unsound", "yanked"]
+
+
+class CargoAuditJSON(TypedDict):
+    """Cargo audit JSON."""
+
+    vulnerabilities: dict[Literal["list"], list[CargoAuditWarning]]
+    warnings: dict[Literal["unsound", "yanked"], list[CargoAuditWarning]]
+
+
 class CargoAuditSecurityWarnings(JSONFileSourceCollector):
     """Cargo Audit collector for security warnings."""
 
     def _parse_json(self, json: JSON, filename: str) -> Entities:
         """Override to parse the security warnings from the JSON."""
         entities = Entities()
-        vulnerabilities = cast(JSONDict, json).get("vulnerabilities", {}).get("list", [])
-        for vulnerability in vulnerabilities:
-            package_name = vulnerability["package"]["name"]
-            package_version = vulnerability["package"]["version"]
-            advisory_id = vulnerability["advisory"]["id"]
+        for warning in self.__warnings(json):
+            package_name = warning["package"]["name"]
+            package_version = warning["package"]["version"]
+            advisory = warning.get("advisory", {}) or {}
+            versions = warning.get("versions", {}) or {}
+            advisory_id = advisory.get("id", "")
             entities.append(
                 Entity(
                     key=f"{package_name}:{package_version}:{advisory_id}",
                     package_name=package_name,
                     package_version=package_version,
                     advisory_id=advisory_id,
-                    advisory_title=vulnerability["advisory"]["title"],
-                    advisory_url=vulnerability["advisory"]["url"],
-                    versions_patched=", ".join(vulnerability.get("versions", {}).get("patched", [])),
+                    advisory_title=advisory.get("title", ""),
+                    advisory_url=advisory.get("url", ""),
+                    versions_patched=", ".join(versions.get("patched", [])),
+                    warning_type=warning.get("kind", "vulnerability"),
                 ),
             )
         return entities
+
+    def __warnings(self, json: JSON) -> list[CargoAuditWarning]:
+        """Collect all warnings into one list for easier processing."""
+        json_dict = cast(CargoAuditJSON, json)
+        warnings = json_dict.get("vulnerabilities", {}).get("list", [])
```

Comment:
Why use `.get()`, shouldn't the fact that it is an instance of `CargoAuditJSON` already mean that these values exist? Or does it secretly classify fields as `Optional`?

Reply:
The data is not validated against the TypedDict like when using Pydantic. It's only type hints. And there's no schema for the Cargo Audit JSON so I don't know for sure these fields are always present.